### DISCUSSION TITLE
close open system ports

### DISCUSSION THREAD

**Discussion by HamzaMiz** [appium/appium#16280] [Q&A]
is there any solution to close all open systemports programmatically before starting appium driver   ?

**Comment by KazuCocoa** [comment]
For example, what ports remained?

Ports used by an appium server will be closed when the process gets SIGINT or SIGTERM as delete existing sessions.
https://github.com/appium/appium/blob/1e30207ec4e413c64396420fbb0388392e88cc54/lib/main.js#L163

So, if you close existing sessions by `delete` WebDriver command or close the appium session via SIGINT/SIGTERM, they will be closed as current implementation.

**Reply by HamzaMiz** [reply]
it says uiautomator2 cannot start because port 8202 is busy

**Reply by KazuCocoa** [reply]
Could you share the log?
AFAIK, the situation might happen when an appium server gets SIGKILL while it has running sessions.

**Reply by Hamzastreamwide** [reply]
[debug] [W3C] Calling AppiumDriver.createSession() with args: [{"app":"/var/lib/jenkins/workspace/SmartMS-Android-GeneratePackage/AndroidPackage/TOTR.apk","appPackage":"com.teamontherun.preprod","allowTestPackages":true,"noReset":true,"clearDeviceLogsOnStart":true,"deviceName":"Android","appWaitDuration":30000,"appActivity":"com.streamwide.smartms.ui.activity.MainActivity","androidInstallTimeout":500000,"adbExecTimeout":30000,"automationName":"UiAutomator2","platformName":"android","udid":"42002add4fe034d3","systemPort":"8202"},null,{"firstMatch":[{"appium:adbExecTimeout":30000,"appium:allowTestPackages":true,"appium:androidInstallTimeout":500000,"appium:app":"/var/lib/jenkins/workspace/SmartMS-Android-GeneratePackage/AndroidPackage/TOTR.apk","appium:appActivity":"com.streamwide.smartms.ui.activity.MainActivity","appium:appPackage":"com.teamontherun.preprod","appium:appWaitDuration":30000,"appium:automationName":"UiAutomator2","clearDeviceLogsOnStart":true,"appium:deviceName":"Android","appium:noReset":true,"platformName":"android","appium:systemPort":"8202","appium:...
[debug] [BaseDriver] Event 'newSessionRequested' logged at 1640623870703 (17:51:10 GMT+0100 (Central European Standard Time))
[BaseDriver] The following capabilities are not standard capabilities and should have an extension prefix:
[BaseDriver]   clearDeviceLogsOnStart
[Appium] Appium v1.22.2 creating new AndroidUiautomator2Driver (v1.70.1) session
[debug] [BaseDriver] W3C capabilities and MJSONWP desired capabilities were provided
[debug] [BaseDriver] Creating session with W3C capabilities: {
[debug] [BaseDriver]   "alwaysMatch": {
[debug] [BaseDriver]     "appium:clearDeviceLogsOnStart": true,
[debug] [BaseDriver]     "platformName": "android",
[debug] [BaseDriver]     "appium:adbExecTimeout": 30000,
[debug] [BaseDriver]     "appium:allowTestPackages": true,
[debug] [BaseDriver]     "appium:androidInstallTimeout": 500000,
[debug] [BaseDriver]     "appium:app": "/var/lib/jenkins/workspace/SmartMS-Android-GeneratePackage/AndroidPackage/TOTR.apk",
[debug] [BaseDriver]     "appium:appActivity": "com.streamwide.smartms.ui.activity.MainActivity",
[debug] [BaseDriver]     "appium:appPackage": "com.teamontherun.preprod",
[debug] [BaseDriver]     "appium:appWaitDuration": 30000,
[debug] [BaseDriver]     "appium:automationName": "UiAutomator2",
[debug] [BaseDriver]     "appium:deviceName": "Android",
[debug] [BaseDriver]     "appium:noReset": true,
[debug] [BaseDriver]     "appium:systemPort": "8202",
[debug] [BaseDriver]     "appium:udid": "42002add4fe034d3"
[debug] [BaseDriver]   },
[debug] [BaseDriver]   "firstMatch": [
[debug] [BaseDriver]     {}
[debug] [BaseDriver]   ]
[debug] [BaseDriver] }
[BaseDriver] Number capability passed in as string. Functionality may be compromised.
[BaseDriver] Capability 'systemPort' changed from string ('8202') to integer (8202). This may cause unexpected behavior
[BaseDriver] Session created with session id: 7000552b-f4a9-4059-b339-9ac5c70de316
[BaseDriver] Using local app '/var/lib/jenkins/workspace/SmartMS-Android-GeneratePackage/AndroidPackage/TOTR.apk'
[debug] [UiAutomator2] Checking whether app is actually present
[ADB] Found 3 'build-tools' folders under '/home/jenkins/Android/Sdk' (newest first):
[ADB]     /home/jenkins/Android/Sdk/build-tools/32.0.0
[ADB]     /home/jenkins/Android/Sdk/build-tools/31.0.0
[ADB]     /home/jenkins/Android/Sdk/build-tools/29.0.2
[ADB] Using 'adb' from '/home/jenkins/Android/Sdk/platform-tools/adb'
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 start-server'
[AndroidDriver] Retrieving device list
[debug] [ADB] Trying to find a connected android device
[debug] [ADB] Getting connected devices
[debug] [ADB] Connected devices: [{"udid":"42002add4fe034d3","state":"device"},{"udid":"PNXGAM8932703079","state":"device"},{"udid":"R38M40E72YE","state":"device"}]
[AndroidDriver] Using device: 42002add4fe034d3
[ADB] Using 'adb' from '/home/jenkins/Android/Sdk/platform-tools/adb'
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 start-server'
[debug] [ADB] Setting device id to 42002add4fe034d3
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 shell getprop ro.build.version.sdk'
[debug] [ADB] Current device property 'ro.build.version.sdk': 28
[ADB] Getting device platform version
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 shell getprop ro.build.version.release'
[debug] [ADB] Current device property 'ro.build.version.release': 9
[debug] [ADB] Device API level: 28
[UiAutomator2] Relaxing hidden api policy
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 shell 'settings put global hidden_api_policy_pre_p_apps 1;settings put global hidden_api_policy_p_apps 1;settings put global hidden_api_policy 1''
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 wait-for-device'
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 shell echo ping'
[debug] [AndroidDriver] Pushing settings apk to device...
[debug] [ADB] Getting install status for io.appium.settings
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 shell dumpsys package io.appium.settings'
[debug] [ADB] 'io.appium.settings' is installed
[debug] [ADB] Getting package info for 'io.appium.settings'
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 shell dumpsys package io.appium.settings'
[debug] [ADB] The version code of the installed 'io.appium.settings' is greater than the application version code (31 > 30)
[debug] [ADB] There is no need to downgrade 'io.appium.settings'
[debug] [ADB] Getting IDs of all 'io.appium.settings' processes
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 shell 'pgrep --help; echo $?''
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 shell pgrep -f \(\[\[:blank:\]\]\|\^\)io\.appium\.settings\(\[\[:blank:\]\]\|\$\)'
[debug] [AndroidDriver] io.appium.settings is already running. There is no need to reset its permissions.
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 shell appops set io.appium.settings android:mock_location allow'
[debug] [Logcat] Clearing logcat logs from device
[debug] [Logcat] Starting logs capture with command: /home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 logcat -v threadtime
[debug] [UiAutomator2] Forwarding UiAutomator2 Server port 6790 to local port 8202
[UiAutomator2] UiAutomator2 Server cannot start because the local port #8202 is busy. Make sure the port you provide via 'systemPort' capability is not occupied. This situation might often be a result of an inaccurate sessions management, e.g. old automation sessions on the same device must always be closed before starting new ones.
[debug] [UiAutomator2] Deleting UiAutomator2 session
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 shell am force-stop com.teamontherun.preprod'
[debug] [Logcat] Stopping logcat capture
[debug] [ADB] Removing forwarded port socket connection: 8202 
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 forward --remove tcp:8202'
[UiAutomator2] Restoring hidden api policy to the device default configuration
[debug] [ADB] Running '/home/jenkins/Android/Sdk/platform-tools/adb -P 5037 -s 42002add4fe034d3 shell 'settings delete global hidden_api_policy_pre_p_apps;settings delete global hidden_api_policy_p_apps;settings delete global hidden_api_policy''
[debug] [BaseDriver] Event 'newSessionStarted' logged at 1640623872312 (17:51:12 GMT+0100 (Central European Standard Time))
[debug] [W3C] Encountered internal error running command: Error: UiAutomator2 Server cannot start because the local port #8202 is busy. Make sure the port you provide via 'systemPort' capability is not occupied. This situation might often be a result of an inaccurate sessions management, e.g. old automation sessions on the same device must always be closed before starting new ones.
[debug] [W3C]     at Object.errorAndThrow (/home/linuxbrew/.linuxbrew/lib/node_modules/appium/node_modules/appium-support/lib/logging.js:94:35)
[debug] [W3C]     at forwardPort (/home/linuxbrew/.linuxbrew/lib/node_modules/appium/node_modules/appium-uiautomator2-driver/lib/driver.js:284:16)
[debug] [W3C]     at AndroidUiautomator2Driver.allocateSystemPort (/home/linuxbrew/.linuxbrew/lib/node_modules/appium/node_modules/appium-uiautomator2-driver/lib/driver.js:294:14)
[debug] [W3C]     at AndroidUiautomator2Driver.startUiAutomator2Session (/home/linuxbrew/.linuxbrew/lib/node_modules/appium/node_modules/appium-uiautomator2-driver/lib/driver.js:383:5)
[debug] [W3C]     at AndroidUiautomator2Driver.createSession (/home/linuxbrew/.linuxbrew/lib/node_modules/appium/node_modules/appium-uiautomator2-driver/lib/driver.js:229:7)
[debug] [W3C]     at AppiumDriver.createSession (/home/linuxbrew/.linuxbrew/lib/node_modules/appium/lib/appium.js:387:35)
[HTTP] <-- POST /wd/hub/session 500 1674 ms - 1167
[HTTP] 

org.openqa.selenium.SessionNotCreatedException: Unable to create a new remote session. Please check the server log for more details. Original error: An unknown server-side error occurred while processing the command. Original error: UiAutomator2 Server cannot start because the local port #8202 is busy. Make sure the port you provide via 'systemPort' capability is not occupied. This situation might often be a result of an inaccurate sessions management, e.g. old automation sessions on the same device must always be closed before starting new ones.
Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:17:03'
System info: host: 'jenkinssreamwidetn', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '5.4.0-91-generic', java.version: '1.8.0_292'
Driver info: driver.version: AndroidDriver
remote stacktrace: UnknownError: An unknown server-side error occurred while processing the command. Original error: UiAutomator2 Server cannot start because the local port #8202 is busy. Make sure the port you provide via 'systemPort' capability is not occupied. This situation might often be a result of an inaccurate sessions management, e.g. old automation sessions on the same device must always be closed before starting new ones.
    at getResponseForW3CError (/home/linuxbrew/.linuxbrew/lib/node_modules/appium/node_modules/appium-base-driver/lib/protocol/errors.js:804:9)
    at asyncHandler (/home/linuxbrew/.linuxbrew/lib/node_modules/appium/node_modules/appium-base-driver/lib/protocol/protocol.js:380:37)
Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:17:03'
System info: host: 'jenkinssreamwidetn', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '5.4.0-91-generic', java.version: '1.8.0_292'
Driver info: driver.version: AndroidDriver

**Reply by KazuCocoa** [reply]
Did you delete the previous session?
The error happens when the port forward by adb did not closed. Appium tries to close the port in its delete session.

So, the case could happen when an appium process is killed unexpectedly (e.g. sigkill) instead of sigint/sigterm, or the adb command, which appium issues in its delete session steps, does not close the port correctly.

**Reply by Hamzastreamwide** [reply]
by  delete u mean call driver.quit ?

**Reply by KazuCocoa** [reply]
Yes, https://appium.io/docs/en/commands/session/delete/

**Reply by Hamzastreamwide** [reply]
yes @afterMethed and @afterTest and always run , but i have retry mechanism , when the test failed it's registred as skipped not failed , maybe in Onskip listener , i need to add driver quit there

**Comment by Hamzastreamwide** [comment]
i found this solution , but still didnt test it does it make any sense ? https://stackoverflow.com/questions/66422013/how-do-i-make-sure-ports-8200-8299-for-uiautmator2-are-freed-up-after-killing
